### PR TITLE
Use multiple issue templates with forms for new issues.

### PR DESCRIPTION
**Pull request by Zimmi48**
Demo here: https://github.com/Zimmi48/coq/issues/new/choose (feel free to go all the way and create new issues to see how it goes)

I'm opening this PR to get feedback:
- What do you think of this change?
- What should be the categories / templates to choose from?
- What should the forms ask?
- Do we want to make some fields compulsory? (At the moment, no field is compulsory because I feel like this will avoid putting new barriers on people wanting to report issues.)
- Do we want a template for the title as well (e.g. `[Bug]` / `[Wish]`)?
- What external resources (support links) should we add? (We can have multiple.) Should we include Discourse + Zulip + StackOverflow?
- Do we want to keep the possibility of starting without the template? (It seems useful especially for devs and all the more important given that you cannot remove the form template as you would for the old template.)
- What labels do we want to automatically assign? In particular, does it make sense to automatically assign the `needs: triage` label to consider that bugs *must* be reproduced and feature request *must* be evaluated by a developer / regular contributor (who would then remove the `needs: triage` label and possibly add others instead)? I'd be tempted to say yes and that we should more systematically evaluate the priority and the difficulty levels in particular.

cc @coq/contributing-process-maintainers @herbelin @Alizter @tchajed 

Also, with my colleague @jrfaller, we've envisioned that we could automatically keep reproducible bugs current by having the bot and/or a scheduled GitHub Action retry the minimal examples provided and update the issue OP with up-to-date data. This could supersede the `test-suite/bugs/opened` folder. I think that @JasonGross might be interested in contributing to the design / implementation / evaluation of this idea. So is this a reasonable idea? And how should we amend the template to make it possible to reproduce most issues? (E.g. add the possibility to depend on external projects? How so?)

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+labels: ["kind: bug", "needs: triage"]
+body:
+- type: markdown
+  attributes:
+    value: |
+      Thank you for your contribution.
+
+      It is helpful to explain how to reproduce the bug and what the problem is.
+      If you have a small reproducible example, you can use the second field to provide it.
+      Otherwise, please provide a link to a repository, a gist (https://gist.github.com) or drag-and-drop a `.zip` archive in the field below.
```
Comment (SkySkimmer) [inline .github/ISSUE_TEMPLATE/bug_report.yml:12]: "in the field below" is a little unclear, should it be "in the first field"?

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+name: Bug report
+description: Report an unexpected behavior
+labels: ["kind: bug", "needs: triage"]
```
Comment (SkySkimmer) [inline .github/ISSUE_TEMPLATE/bug_report.yml:3]: I assume people with rights on the repo can change the labels in the creation page, is that correct?

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    render: coq
+- type: dropdown
+  attributes:
+    label: Version of Coq where this bug occurs
+    description: |
+      You can get this information by running `coqtop -v`.
+      Feel free to select multiple answers if you can reproduce the bug on several versions of Coq.
+    multiple: true
+    options:
+    - 8.15+alpha
```
Comment (SkySkimmer) [inline .github/ISSUE_TEMPLATE/bug_report.yml:34]:
```suggestion
    - 8.15+alpha (current master)
```

**Comment by JasonGross** [comment]
Regarding design and implementation of opened examples, I think with a few tweaks, the existing bug minimizer infrastructure can do most of the heavy lifting.  In particular, the bug minimizer already has support for making sure that the error message matches the one given in a log file, and there trailing end of this log file is posted in it's response.  Things we need to figure out:
- how to mark a comment as the blessed version of the bug (as opposed to the minimizer running on code that errors with something like "you forgot to install a dependency on opam")
- what to do when the error message changes in minor ways (tweaks to grammar, for example)
- ideally, how to get the full build log off of GH Artifacts, in case the error message is not in the last 8 kb
- how to have the bug minimizer runner report differently when the error is "your file didn't produce an error matching the log".  Ideally this template would be used any time we managed to run the bug minimizer (that is, all of the setup succeeded), but the initial sanity checks of the bug minimizer failed.
- how to trigger the minimizer initially (perhaps this is the job of the bug triager?)

I'm excited about the prospect of the bug minimizer getting run on basically all of the bugs reported, and in having their status automatically checked periodically.  Seems pretty cool

**Review by SkySkimmer — APPROVED** [review]
>Do we want to make some fields compulsory? (At the moment, no field is compulsory because I feel like this will avoid putting new barriers on people wanting to report issues.)

no IMO
Or at most the description field.
Is there a way to just skip the section when not filled in instead of outputting `last good version: not answered`?

>What external resources (support links) should we add? (We can have multiple.) Should we include Discourse + Zulip + StackOverflow?

zulip yes
Who is on SO?

>Do we want to keep the possibility of starting without the template? (It seems useful especially for devs and all the more important given that you cannot remove the form template as you would for the old template.)

Yes

>In particular, does it make sense to automatically assign the needs: triage label 

I think it's nice to have needs:triage, it basically means "no dev has looked at this yet". I'm not sure that it makes sense to assign more meaning to it though.

**Comment by JasonGross** [comment]
> and update the issue OP with up-to-date data.

Oh, I missed this bit.  I guess we could keep the blessed version of the bug in the OP, though this makes it even more important to be able to link to the artifacts due to comment size limits.  I imagine the template could have a field for initial reproducing script and error message, and then the bot will edit this comment to fill in the (not-user-fillable) fields for minimized example (and perhaps build log, though I guess we can just use the error message in place of the build log, as log as we add a "File ... characters ... Error:" header).

Should we re-run minimization from scratch any time someone other than coqbot changes the OP?  (Is it possible to detect changes specifically to the initial script field and trigger on that?)

What do we do for multiple examples of the same underlying bug?  Do we want users to report separate issues, or do we want to have the ability to bless multiple minimal examples?

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 2 above)
Comment (Zimmi48) [reply .github/ISSUE_TEMPLATE/bug_report.yml:3]: Yes.

**Comment by Zimmi48** [comment]
>Is there a way to just skip the section when not filled in instead of outputting `last good version: not answered`?

No, there isn't. I will send feedback about this to GitHub. In the meantime, if that's a bother, we could either: duplicate the template to add a "Regression" category or have coqbot detect and remove these `last good version: not answered` bits by editing the issue right after it's opened.

>  Who is on SO?

I don't follow it actively (lack of time) but all Stack Overflow and Stack Exchange questions about Coq are [cross posted to Zulip](https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/New.20Stack.20Exchange.20question) and when I look at some of them, it seems that they are generally pretty well answered. See here for a list of active answerers: https://stackoverflow.com/tags/coq/topusers

>I think it's nice to have needs:triage, it basically means "no dev has looked at this yet". I'm not sure that it makes sense to assign more meaning to it though.

:+1: though we should still document the minimal expectation for removing the `needs: triage` label (in the contributing guide). Otherwise, devs will look but will still hesitate to remove the label. In particular, it should be clarified that it's OK to remove the label without commenting.

**Comment by Zimmi48** [comment]
> Regarding design and implementation of opened examples, I think with a few tweaks, the existing bug minimizer infrastructure can do most of the heavy lifting. In particular, the bug minimizer already has support for making sure that the error message matches the one given in a log file, and there trailing end of this log file is posted in it's response.

You mean that the "bug minimizer" is still useful even if there is nothing to minimize, but just for making sure that the error remains (mostly) the same across versions of Coq?

What about then adding a separate field where the user is supposed to copy-paste the user message (be it an error message, a warning or a weird feedback) that they get?

If not provided, we could also consider the following process:
- coqbot posts a comment asking if the error message it detected is the expected one;
- the user can validate this by reacting with a :+1: (no need to post a comment);
- then coqbot edits the OP with the additional info.

>what to do when the error message changes in minor ways (tweaks to grammar, for example)

What about coqbot posting a comment asking a dev to validate that the new message still reflects the same bug (same validation method with a silent :+1: then coqbot edits the OP)?

>how to have the bug minimizer runner report differently when the error is "your file didn't produce an error matching the log". 

What about coqbot posting a comment informing of the error + adding a label that it's "not automatically reproducible"?

>edit this comment to fill in the (not-user-fillable) fields for minimized example

What about overriding the initial example with the new minimized version if we do produce a more minimal version?

>Is it possible to detect changes specifically to the initial script field and trigger on that?

I think that we do get precise information of what was before and after when an issue is edited.

>What do we do for multiple examples of the same underlying bug? Do we want users to report separate issues, or do we want to have the ability to bless multiple minimal examples?

Good question. We could suggest separating multiple examples with marker comments such as `(* example X *)` and matching `(* output X *)`. Unfortunately, GitHub issue forms do not support advance workflows like asking if there are more examples to add and inserting a new code textarea.

**Comment by Alizter** [comment]
Since you have a link to the discourse forum, would it also be useful to have a link to Zulip?

**Comment by Zimmi48** [comment]
Indeed, that was one of my questions:

>What external resources (support links) should we add? (We can have multiple.) Should we include Discourse + Zulip + StackOverflow?

**Comment by Alizter** [comment]
I don't know if we want to *encourage* posting on stackoverflow since it is somewhat likely the question will be missed. I am also aware that we have a SO bot on Zulip. I think at least Zulip would be good to include.

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+name: Bug report
+description: Report an unexpected behavior
```
Comment (Alizter) [inline .github/ISSUE_TEMPLATE/bug_report.yml:2]: Do we want UK or US English spelling?

Comment (Alizter) [reply .github/ISSUE_TEMPLATE/bug_report.yml:2]:
```suggestion
description: Report an unexpected behavior.
```

Comment (Zimmi48) [reply .github/ISSUE_TEMPLATE/bug_report.yml:2]: US English spelling everywhere.

**Comment by Zimmi48** [comment]
>it is somewhat likely the question will be missed

What do you mean by this exactly?

Anyway, SO does provide some statistics on the `coq` tag:

Number of questions|Period|Unanswered
-- | -- | --
3 | Last 7 Days | 0% unanswered
37 | Last 30 Days | 24.3% unanswered
2,464 | All Time | 10.3% unanswered

I can't compare them with Discourse and/or Zulip because I don't know how to extract this kind of stats for those, but I regularly see unanswered questions on Discourse as well.

**Comment by Zimmi48** [comment]
In terms of level of activity, Discourse receives in fact less new questions than Stack Overflow it would seem:

  | Last 24 hours | Last 7 days | Last 30 days | All Time
-- | -- | -- | -- | --
Topics | 0 | 2 | 12 | 541
Posts | 0 | 3 | 65 | 4.0k

I know that Zulip receives a lot, but I wish it would be less used since it's not as good for long-term archival / retrieval.

**Comment by Alizter** [comment]
Sorry, I don't know what I am talking about with regards to SO so ignore that.

The (I think about:?) descriptions on two of the templates are missing full stops. Also should the field titles in the templates end with full stops? At the moment we have both.

**Comment by Zimmi48** [comment]
>The (I think about:?) descriptions on two of the templates are missing full stops.

Fixed, thanks!

>Also should the field titles in the templates end with full stops? At the moment we have both.

Indeed, I'm being inconsistent. I would tend to not put them, but I dunno.

**Comment by Alizter** [comment]
I think not putting them looks more inviting.

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    - 8.13.0
+    - 8.12.2
+    - 8.12.1
+    - 8.12.0
+    - 8.11.2
+    - 8.11.1
+    - 8.11.0
+- type: dropdown
+  attributes:
+    label: Last version of Coq where the bug did not occur.
```
Comment (Alizter) [inline .github/ISSUE_TEMPLATE/bug_report.yml:47]:
```suggestion
    label: Last version of Coq where the bug did not occur
```

**Review by Alizter — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Any news? If this is dead maybe it should be closed.

**Comment by Zimmi48** [comment]
I think all the people who looked at this liked it, and looking at it almost a year after, I still like it. Let's discuss it at the next Coq Call maybe?

**Review by Alizter — APPROVED** [review]

**Comment by SkySkimmer** [comment]
The coq call notes said https://github.com/coq/coq/wiki/Coq-Call-2022-10-19
>Issue forms sound like a good idea. Minor refinements to make. But will need to be discussed again when ready to merge since we were only four (Enrico, Gaëtan, Pierre-Marie, Théo).

I don't remember what minor refinements were talked about. Should we schedule in a new call or just merge?

**Comment by SkySkimmer** [comment]
added to next call

**Comment by Zimmi48** [comment]
Here is what my own notes from this call said:

>Add Zulip and ProofAssistants SE. Don't put a scrolling menu for Coq versions.

**Comment by Zimmi48** [comment]
And from what I can tell from the PR timeline, I did implement these changes right after the call.

**Comment by SkySkimmer** [comment]
Discussion on zulip points towards merging now.

@coqbot merge now

**Pull request merged**